src/sentry/sentry_metrics/indexer/limiters/writes.py

```diff
@@ -109,80 +51,172 @@ def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
         """
         Consumes the rate limits returned by `check_write_limits`.
         """
-        if exc_type is None:
-            self._writes_limiter.rate_limiter.use_quotas(
-                self._requests, self._grants, self._timestamp
-            )
+        if exc_type is not None:
+            return
+        for requests, grants, timestamp in zip(self._requests, self._grants, self._timestamps):
+            self._writes_limiter.rate_limiter.use_quotas(requests, grants, timestamp)
 
 
 class WritesLimiter:
     def __init__(self, namespace: str, **options: Mapping[str, str]) -> None:
         self.namespace = namespace
         self.rate_limiter: RedisSlidingWindowRateLimiter = RedisSlidingWindowRateLimiter(**options)
 
+    def _build_quota_key(self, org_id: Optional[OrgId] = None) -> str:
+        if org_id is not None:
+            return f"metrics-indexer-{self.namespace}-org-{org_id}"
+        else:
+            return f"metrics-indexer-{self.namespace}-global"
+
+    @metrics.wraps("sentry_metrics.indexer.construct_quotas")
+    def _construct_quotas(self, use_case_id: UseCaseID) -> Sequence[Quota]:
+        """
+        Construct write limit's quotas based on current sentry options.
+
+        This value can potentially cached globally as long as it is invalidated
+        when sentry.options are.
+        """
+        if use_case_id is UseCaseID.TRANSACTIONS:
+            return [
+                Quota(prefix_override=self._build_quota_key(), **args)
+                for args in options.get("sentry-metrics.writes-limiter.limits.performance.global")
+            ] + [
+                Quota(prefix_override=None, **args)
+                for args in options.get("sentry-metrics.writes-limiter.limits.performance.per-org")
+            ]
+        elif use_case_id is UseCaseID.SPANS:
+            return [
+                Quota(prefix_override=self._build_quota_key(), **args)
+                for args in options.get("sentry-metrics.writes-limiter.limits.spans.global")
+            ] + [
+                Quota(prefix_override=None, **args)
+                for args in options.get("sentry-metrics.writes-limiter.limits.spans.per-org")
+            ]
+        elif use_case_id is UseCaseID.SESSIONS:
```

Comment:
Although checking for each of the 3 use cases is short-ish for now, can we maybe create a mapping to simplify the logic here? (At some point, we may have many use cases residing on the performance pipeline, and we don't want to check for each individually)

Reply:
Okay, refactored the the use case ID to option name mapping into `use_case_id_registry.py`